### PR TITLE
New: eslint-visitor-keys

### PR DESCRIPTION
**Pull request by mysticatea**
This PR is the first implementation of `eslint-visitor-keys`.

Please review for direction.

Also, I'm not sure how I can setup scripts to release this package.
Please help me.

### CODE REVIEW

**Comment by ilyavolodin** [comment]
Please don't merge this PR in yet. JSF prefers Apache license to MIT. So while we still don't have any code in the repository, it would be easier to change the license.

File: README.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+Get the visitor keys of a given AST node.
+
+This is similar to `Object.keys(node)` of ES Standard, but some keys are excluded: `parent`, `leadingComments`, `trailingComments`, and names which start with `_`.
+
+This will be used to traverse unknown nodes.
+
+For example:
+
+```
+console.log(evk.getKeys()) // → ["left", "right"]
```
Comment (not-an-aardvark) [inline README.md]: Should this call have an argument?

File: lib/index.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
+/**
+ * @author Toru Nagashima <https://github.com/mysticatea>
+ * See LICENSE file in root directory for full license.
+ */
+"use strict";
+
+const KEYS = require("./visitor-keys.json");
+
+// Types.
+const NODE_TYPES = Object.freeze(Object.keys(KEYS));
```
Comment (not-an-aardvark) [inline lib/index.js]: Does `NODE_TYPES` need to be frozen? It seems like it's not exported.

(I suppose it's not a problem to freeze it either way.)

File: tests/lib/index.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            for (const type of Object.keys(additionalKeys)) {
+                for (const key of additionalKeys[type]) {
+                    assert(unionKeys[type].indexOf(key) !== -1, `'${key}' should be included in '${type}'.`);
+                }
+            }
+        });
+
+        it("should not have duplicate", () => {
+            assert(unionKeys.Program.filter(key => key === "body").length === 1);
+        });
```
Comment (not-an-aardvark) [inline tests/lib/index.js]: Should we add tests for preserving the order of keys?

```js
evk.unionWith({ CallExpression: ["callee"] }).CallExpression;

// ["callee", "arguments"]?
// ["arguments", "callee"]?
```

File: README.md (hunk 1 above)
Comment (mysticatea) [reply README.md]: Oh, yes! This is a copy-paste mistake.

File: lib/index.js (hunk 2 above)
Comment (mysticatea) [reply lib/index.js]: Because of my liking. It clarifies that it's a constant.

**Comment by mysticatea** [comment]
I update this PR to follow the review.

I rethought about the order of keys for `unionWith()`. Users need to control the order to traverse nodes from fore to back (e.g. `decorators` → `key` → `value`).

**Comment by JamesHenry** [comment]
Thanks for driving this, @mysticatea! I'll use this PR commit in my babel-eslint PR and see where we're at

**Comment by JamesHenry** [comment]
@mysticatea Your freezing of the keys was very useful in flagging up a remaining case of mutation happening within babel-eslint.

The contents of this function are unchanged in my PR, but the file it is now in is different:

https://github.com/babel/babel-eslint/blob/64bd56509e41b31131ff08db882b261deb34d62d/lib/enhanced-referencer.js#L76-L110

```js
// set ArrayPattern/ObjectPattern visitor keys back to their original. otherwise
    // eslint-scope will traverse into them and include the identifiers within as declarations
    extendedVisitorKeys.ObjectPattern = ["properties"];
    extendedVisitorKeys.ArrayPattern = ["elements"];
    // Execute default Referencer logic
    super.visitFunction(node);
    // set them back to normal...
    extendedVisitorKeys.ObjectPattern = t.VISITOR_KEYS.ObjectPattern;
    extendedVisitorKeys.ArrayPattern = t.VISITOR_KEYS.ArrayPattern;
```

What do you think about this? Should we try and expose an API in this new project for temporarily overriding keys adhoc like this? Or is there a better way?

**Comment by mysticatea** [comment]
@JamesHenry Does it behave differently about patterns between inside of functions and outside of functions? I think `super.visitFunction(node);` visits the body of functions.

```js
const {a} = obj // original visitor keys.

function foo() {
    const {a} = obj // overwritten visitor keys.
}

**Comment by JamesHenry** [comment]
I don't know, I'm afraid. This PR is my first interaction with babel-eslint, and (eslint-scope's Referencer in general), and as I say this is not my code...

**Comment by JamesHenry** [comment]
I will have much more time next week for open-source, I am a bit swamped right now

**Comment by mysticatea** [comment]
Thanks. I think that we don't support temporarily overriding for keys in this repo. Traversing is recursive, so it causes unintentional behaviors in general.

**Comment by mysticatea** [comment]
May I merge this?

**Comment by not-an-aardvark** [comment]
:+1:

**Pull request merged**